src/adapter/src/optimize/index.rs

```diff
@@ -8,21 +8,33 @@
 // by the Apache License, Version 2.0.
 
 //! Optimizer implementation for `CREATE INDEX` statements.
+//!
+//! Note that, in contrast to other optimization pipelines, timestamp selection is not part of
+//! index optimization. Instead users are expected to separately set the as-of on the optimized
+//! `DataflowDescription` received from `GlobalLirPlan::unapply`. Reasons for choosing to exclude
+//! timestamp selection from the index optimization pipeline are:
+//!
+//!  (a) Indexes don't support non-empty `until` frontiers, so they don't provide opportunity for
+//!      optimizations based on the selected timestamp.
```

Comment:
I realize that this is lying a bit. We can apply single-timestamp optimizations when we select the as-of to be `u64::MAX`, for which the empty `until` frontier would be considered "one after that". That happens for indexes on constant queries.

However, I'm not sure we have to care about that. Indexes (or MVs) on constant queries seem quite useless, and it's unlikely for anyone to even create complex views that would benefit from single-timestamp optimizations on only constant data. Still this is something we (I, at least) didn't consider before.

I'm fine with keeping it like this for the reasons mentioned above, but what do you think @aalexandrov? The alternative would be to move back to having the timestamp resolution on the `GlobalMirPlan` and only partially optimize plans during bootstrapping.

Reply:
I don't think this warrants the extra complexity by itself. We just need to somehow keep track of the reasons to go back to the current state.